File: src/main/java/org/strassburger/lifestealz/listeners/PlayerLoginListener.java
package org.strassburger.lifestealz.listeners;

import net.kyori.adventure.text.Component;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerLoginEvent;
import org.strassburger.lifestealz.LifeStealZ;
import org.strassburger.lifestealz.util.MessageUtils;
import org.strassburger.lifestealz.util.WhitelistManager;
import org.strassburger.lifestealz.storage.PlayerData;
import org.strassburger.lifestealz.storage.Storage;

public final class PlayerLoginListener implements Listener {
    private final LifeStealZ plugin;

    public PlayerLoginListener(LifeStealZ plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onPlayerLogin(PlayerLoginEvent event) {
        Player player = event.getPlayer();
        Storage storage = plugin.getStorage();

        if (!WhitelistManager.isWorldWhitelisted(player)) return;

        PlayerData playerData = loadOrCreatePlayerData(player, storage);

        if (shouldKickPlayer(playerData)) {
            kickPlayer(event);
        }
    }

    private PlayerData loadOrCreatePlayerData(Player player, Storage storage) {
        PlayerData playerData = plugin.getStorage().load(player.getUniqueId());
        if (playerData == null) {
            playerData = new PlayerData(player.getName(), player.getUniqueId());
            playerData.setFirstJoin(System.currentTimeMillis());
            storage.save(playerData);
            plugin.getGracePeriodManager().startGracePeriod(player);
            plugin.getOfflinePlayerCache().addItem(player.getName());
        }
        return playerData;
    }

    private boolean shouldKickPlayer(PlayerData playerData) {
        boolean disabledBanOnDeath = plugin.getConfig().getBoolean("disablePlayerBanOnElimination");
        double minHearts = plugin.getConfig().getInt("minHearts") * 2;
        return playerData.getMaxHealth() <= minHearts && !disabledBanOnDeath;
    }

    private void kickPlayer(PlayerLoginEvent event) {
        event.setResult(PlayerLoginEvent.Result.KICK_OTHER);
        Component kickmsg = MessageUtils.getAndFormatMsg(false, "eliminatedJoin", "&cYou don't have any hearts left!");
        event.kickMessage(kickmsg);
    }
}

File: src/main/java/org/strassburger/lifestealz/listener/PlayerLoginListener.kt
package org.strassburger.lifestealz.listener

import net.kyori.adventure.text.Component
import org.bukkit.BanEntry
import org.bukkit.BanList
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerLoginEvent
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.scheduler.BukkitRunnable
import org.strassburger.lifestealz.Lifestealz
import org.strassburger.lifestealz.util.ManagePlayerdata

class PlayerLoginListener(private val plugin: JavaPlugin) : Listener {
    private val invulnerabilityDuration = 20L
    private val disabledBanOnDeath = Lifestealz.instance.config.getBoolean("disablePlayerBanOnElimination")

    @EventHandler
    fun playerLoginFunction(event: PlayerLoginEvent) {
        val player = event.player

        val playerData = ManagePlayerdata().getPlayerData(name = player.name, uuid = player.uniqueId.toString())

        applyInvulnerability(player)

        val banList: BanList = Bukkit.getBanList(BanList.Type.NAME)

        if (banList.isBanned(player.name)) {
            val banEntry: BanEntry? = banList.getBanEntry(player.name)
            if (banEntry != null) {
                event.result = PlayerLoginEvent.Result.KICK_BANNED
                val reason = banEntry.reason?: "none"
                event.kickMessage(Component.text("§cYou have been banned for the following reason:\n\n§r$reason"))
            }
            return
        }

        if (playerData.maxhp <= 0.0 && !disabledBanOnDeath) {

            event.result = PlayerLoginEvent.Result.KICK_OTHER
            val kickmsg = Lifestealz.formatMsg(false, "messages.eliminatedjoin", "&cYou don't have any hearts left!")
            event.kickMessage(Component.text(kickmsg))

        }
    }

    private fun applyInvulnerability(player: Player) {
        // Set player invulnerable
        player.isInvulnerable = true

        // Schedule a task to remove invulnerability after the specified duration
        object : BukkitRunnable() {
            override fun run() {
                player.isInvulnerable = false
            }
        }.runTaskLater(plugin, invulnerabilityDuration)
    }
}